### ISSUE TITLE
Implement PollyCache using a action filter

### ISSUE THREAD

**Issue by umeshkarthik** [App-vNext/Polly#521]
I have been trying to implement PolyCAche for my controller action method as a action filter which basically uses the in-memory caching mechanism

**Comment by reisenberger** [comment]
I have an outline for how this could work. In the next few days I will aim to post (here or in the similar question in our slack channel), a minimal example, which you or the community may be able to build further on.

**Comment by reisenberger** [comment]
@umeshkarthik I have an outline for how this _could_ work, but there are some significant potential caveats.

The main concern is that an action-filter solution would have to cache items of type `IActionResult` - but only some implementations of `IActionResult` represent static data which it makes sense to cache.

+ `FileContentResult` and `ContentResult` do (for example) [hold static data](https://github.com/aspnet/Mvc/blob/release/2.1/src/Microsoft.AspNetCore.Mvc.Core/ContentResult.cs#L14-L27).  
+ Other implementations of `IActionResult` do not hold static data: for example [`FileStreamResult`](https://github.com/aspnet/AspNetWebStack/blob/749384689e027a2fcd29eb79a9137b94cea611a8/src/System.Web.Mvc/FileStreamResult.cs).  

_Background_: The return type of a Controller method is an [`IActionResult`](https://github.com/aspnet/Mvc/blob/release/2.1/src/Microsoft.AspNetCore.Mvc.Abstractions/IActionResult.cs) and this is also the type [that action filters must set when they short-circuit the execution](https://docs.microsoft.com/en-us/aspnet/core/mvc/controllers/filters?view=aspnetcore-2.1#cancellation-and-short-circuiting). So the cache would have to cache (and return, on cache hit) items of type `IActionResult`.  But `IActionResult` implementations are in principle dynamic and would be [_reexecuted_ to generate the result](https://github.com/aspnet/Mvc/blob/release/2.1/src/Microsoft.AspNetCore.Mvc.Abstractions/IActionResult.cs#L20), each time they are served from cache.

To illustrate some of the potential risks with `FileStreamResult` ... I _have not tried this yet in practice_ but a number of concerns come to mind with the idea of caching (and re-serving a previously cached) `FileStreamResult`:

+ If it worked (functionally), it would in any case be inefficient, as re-serving the file from the file stream would [re-read the file stream](https://github.com/aspnet/Mvc/blob/a67d9363e22be8ef63a1a62539991e1da3a6e30e/src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/FileResultExecutorBase.cs#L368-L376)
+ Some streams are forward-only, and may not be amenable to being read a second time.
+ Caching something such as a `FileStreamResult` could constitute a resource leak - `Stream` is an `IDisposable` that is probably not intended to be long-lived in a cache.

Other `IActionResult` implementations which might at first glance seem innocuous to cache also have minor issues.  For example, a `JSonResult` might be thought to be static data which is sensible to cache.  But caching a [`JSonResult`](https://github.com/aspnet/Mvc/blob/eda647e27012cb02bab3474df2c093c39261c90e/src/Microsoft.AspNetCore.Mvc.Formatters.Json/JsonResult.cs) would actually cache [the raw underlying object](https://github.com/aspnet/Mvc/blob/eda647e27012cb02bab3474df2c093c39261c90e/src/Microsoft.AspNetCore.Mvc.Formatters.Json/JsonResult.cs#L58-L61) (which might be unexpected or less efficient than caching the JSON string in some cases); and the [json serialization would be re-executed on each serving from cache](https://github.com/aspnet/Mvc/blob/eda647e27012cb02bab3474df2c093c39261c90e/src/Microsoft.AspNetCore.Mvc.Formatters.Json/Internal/JsonResultExecutor.cs#L123-L130).

**TL;DR** This does not mean an action-filter solution cannot be delivered; but it might only make sense for (might be best limited to) a small subset of static content types such as `FileContentResult` and `ContentResult`. Or if not, it would have to be used with deep understanding and care.

@umeshkarthik It would be useful if you could post or state examples (eg method signatures) of the kinds of Controller methods you were thinking of decorating with an action filter.  And, given the above, is this still useful to you?

**Comment by umeshkarthik** [comment]
Hi Dylan

Thank you for the response. Fortunately I was able to achieve the desired
functionality. Now I was into a situation where I need to serialize the
result into a json format before we write it to the cache. Can we override
the OnCacheput method to do that?
Thanks
Umesh K

**Comment by reisenberger** [comment]
@umeshkarthik If you have an implementation with Polly CachePolicy around action filters we would love to see it and potentially share that with the community. I spent some time working on this following your request, and there were some interesting problems.

Re:

>I need to serialize the result into a json format

Please see the [documentation in the Polly wiki on serialization](https://github.com/App-vNext/Polly/wiki/Cache#working-with-serializers) and the [Polly.Caching.Serialization.Json](https://github.com/App-vNext/Polly.Caching.Serialization.Json) library.

**Comment by umeshkarthik** [comment]
Hi Dylan,
My approach was that i created an action filter like this

public class CacheFilter: ActionFilterAttribute, IActionFilter
    {
        private ICacheService _service;
        private TtlStrategy _ttlstartegy { get; set; }
        private KeyStrategy _keystartegy { get; set; }
        private double _timespan { get; set; }

        public CacheFilter(ICacheService service, TtlStrategy ttlstartegy,
KeyStrategy keystartegy, double timespan)
        {
            _service = service;
            _ttlstartegy = ttlstartegy;
            _keystartegy = keystartegy;
            _timespan = timespan;
        }

        public override async Task
OnActionExecutionAsync(ActionExecutingContext context,
ActionExecutionDelegate next)
        {
            var response = _service.Execute(_ttlstartegy, _keystartegy,
TimeSpan.FromMinutes(_timespan), context.HttpContext.Request, next);
            context.Result = response;

        }

Now on the execute I have this

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Frontpoint.Ecommerce.Api.Core.Enums;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.Mvc.Filters;
using Newtonsoft.Json;
using Polly;
using Polly.Caching;
using Polly.Registry;

namespace Frontpoint.Ecommerce.Api.Core
{

    public class MemoryCache : ICacheService
    {
        Polly.Caching.Memory.MemoryCacheProvider _memoryCacheProvider
     = new Polly.Caching.Memory.MemoryCacheProvider(new
Microsoft.Extensions.Caching.Memory.MemoryCache(new
Microsoft.Extensions.Caching.Memory.MemoryCacheOptions()));
        public IActionResult Execute(TtlStrategy ttlStartegy, KeyStrategy
keyStartegy, TimeSpan ts, HttpRequest request, ActionExecutionDelegate next)
        {
            var serializerSettings = new JsonSerializerSettings()
            {
                // Any configuration options
            };

            CachePolicy<IActionResult> cachePolicy = null;
            ITtlStrategy ttl;
            Context policyexecutionContext;
            if (ttlStartegy == TtlStrategy.Relative)
            {
                ttl = new RelativeTtl(ts);
                cachePolicy =
Policy.Cache(_memoryCacheProvider.For<IActionResult>(), ttl);

            }
            else if (ttlStartegy == TtlStrategy.Sliding)
            {
                ttl = new SlidingTtl(ts);
                cachePolicy =
Policy.Cache(_memoryCacheProvider.For<IActionResult>(), ttl);
            }
            else if (ttlStartegy == TtlStrategy.OkResult)
            {

                Func<Context, IActionResult, Ttl> cacheOnly200OKfilter =
        (context, result) => new Ttl(
            timeSpan:
((Microsoft.AspNetCore.Mvc.ObjectResult)result).StatusCode == 200 ? ts :
TimeSpan.Zero,
            slidingExpiration: true
        );
                cachePolicy =
Policy.Cache(_memoryCacheProvider.For<IActionResult>(), new
ResultTtl<IActionResult>(cacheOnly200OKfilter),);
            }

            else if (ttlStartegy == TtlStrategy.Contextual)
            {
                ttl = new ContextualTtl();
                cachePolicy =
Policy.Cache(_memoryCacheProvider.For<IActionResult>(), ttl);

            }

            if (keyStartegy == KeyStrategy.Get)
            {
                policyexecutionContext = new
Context(request.Url().CreateMD5());
            }
            else
            {
                policyexecutionContext = new Context(request.GetPostBody());
            }
            return cachePolicy.Execute(_context => { return
next.Invoke().Result.Result; }, policyexecutionContext);
        }
    }
}

**Comment by umeshkarthik** [comment]
Dylan,

Can you please provide me with an approach on serialising my IActionresult
as Json.

Thanks
Umesh K

**Comment by reisenberger** [comment]
Closing this out as I think we covered this and a number of related issues in other posts.  However, please do open another issue if you require further assistance.
